include/deal.II/lac/affine_constraints.templates.h

```diff
@@ -322,20 +322,123 @@ namespace internal
 
     // helper function
     const auto sort_and_make_unique =
-      [](std::vector<ConstraintType> &constraints) {
-        std::sort(constraints.begin(),
-                  constraints.end(),
+      [&constraints_in, &locally_owned_dofs](
+        std::vector<ConstraintType> &locally_relevant_constraints) {
+        if (locally_relevant_constraints.empty())
+          return;
+
+        for (auto &entry : locally_relevant_constraints)
+          std::sort(entry.entries.begin(),
+                    entry.entries.end(),
+                    [](const auto &l1, const auto &l2) {
+                      return l1.first < l2.first;
+                    });
+
+        std::sort(locally_relevant_constraints.begin(),
+                  locally_relevant_constraints.end(),
                   [](const ConstraintType &l1, const ConstraintType &l2) {
                     return l1.index < l2.index;
                   });
 
-        constraints.erase(std::unique(constraints.begin(),
-                                      constraints.end(),
-                                      [](const ConstraintType &l1,
-                                         const ConstraintType &l2) {
-                                        return l1.index == l2.index;
-                                      }),
-                          constraints.end());
+        auto equal_with_tol = [](const number d, const number e) {
+          if (std::abs(std::real(d - e)) <
+              100000. *
```

Comment:
I am not sure about this factor. This looks like to be hard-coded for `double`. But for `float`, this number seems to be too big. What about:

https://github.com/dealii/dealii/blob/d7105ea62aacb1c6ca5ada834c8926c34d120fbd/include/deal.II/lac/tensor_product_matrix.h#L273

Reply:
What do you think about the function from the example here? This is floating point type independent and a relative measure.

https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon

Reply:
Similar things are used in many places of the library (you can grep for `numeric_limits.*epsilon` to see just those on a single line, so we should probably take a systematic approach if we use them. How did we arrive at the 100000 number, by the way?

Reply:
If I remember correctly the general approach of checking is copied from another place in the library. In an experiment (with double precision) we triggered this value when the tolerance was too small.

What about we use @peterrum s suggestion for now and discuss the best practice for checking equal values separately?

Reply:
Updated!